### DISCUSSION TITLE
No way to use Continue.dev in Visual studio with Openwebui

### DISCUSSION THREAD

**Discussion by flefevre** [open-webui/open-webui#3210] [Q&A]
I have tried ti use the Continue.dev extension plugin in Visual Studio described here : [https://docs.openwebui.com/tutorial/continue-dev](https://docs.openwebui.com/tutorial/continue-dev)

I am using Openwebui and Ollama in a docker configuration.
There are well working (thanks for the job done)

But no way to generate any code in visual studio whatever the parameter:
"apiBase": "http://127.0.0.1:8080/ollama/v1",
"apiBase": "http://localhost:8080/ollama/v1",
"apiBase": "http://openwebui:8080/ollama/v1"

I have generated my sk key in openwebui

I do confirm, I can go in ggole chrome at 
http://localhost:8080/ollama/v1

My config.json file
`{
  "models": [
    {
      "title": "codellama:7b",
      "provider": "openai",
      "model": "codellama:7b",
      "apiBase": "http://127.0.0.1:8080/ollama/v1",
      "apiKey": "sk-XXXX"
    }
  ],
  "customCommands": [
    {
      "name": "test",
      "prompt": "{{{ input }}}\n\nWrite a comprehensive set of unit tests for the selected code. It should setup, run tests that check for correctness including important edge cases, and teardown. Ensure that the tests are complete and sophisticated. Give the tests just as chat output, don't edit any file.",
      "description": "Write unit tests for highlighted code"
    }
  ],
  "tabAutocompleteModel": {
    "title": "codellama:7b",
    "provider": "openai",
    "model": "codellama:7b",
    "apiBase": "http://127.0.0.1:8080/ollama/v1",
    "apiKey": "sk-XXX"
  },
  "allowAnonymousTelemetry": false
}`

Typical error messages i got:

`Error streaming diff: Error: HTTP 500 Internal Server Error from http://127.0.0.1:8080/ollama/v1/completions {"detail":"Open WebUI: Server Connection Error"}`

`request to http://openwebui:8080/ollama/v1/completions failed, reason: getaddrinfo EAI_AGAIN openwebui`

Any idea on how to do it?

**Comment by flefevre** [comment]
I have tried again. I am still getting an error

`Error streaming diff: Error: request to http://openwebui:8080/ollama/v1/completions failed, reason: getaddrinfo EAI_AGAIN openwebui`

do i need to expose the 8080 ports?

can i use https://openwebui.dev.localhost/ollama/v1/? if yes, how to authorize selfsigned certificates?

thanks

**Reply by endogh** [reply]
my vs code  +  continue extention is worked.. im follow this  https://docs.openwebui.com/tutorials/integrations/continue-dev/

**Comment by ProjectMoon** [comment]
It is possible to use. I have it set up and it works fine (insofar as the models are good). You need to make sure that your reverse proxy is set up correctly for streaming the ollama API.

The V1 endpoint is ollama's OpenAI compatibility layer. The API base URL should be `https://example.com/ollama`. Provider type should be set to ollama. This will cause Continue to send requests to `https://example.com/ollama/api` endpoints.

Note, if you want to use ollama embeddings for the embeddings provider, you very specifically need a trailing slash in the ollama base URL in the embeddings provider settings.  I think this is a bug in Continue itself.

**Reply by Shadowfita** [reply]
This worked for me, and differs from the documentation on the site.

**Comment by flefevre** [comment]
I do confirm .it works with the given tutorial.

**Comment by jiatao99** [comment]
I cannot get it to work. I can access http://192.168.1.65:3000 from browser, however, I got this error: 
```
HTTP 400 Bad Request
from http://192.168.1.65:3000/ollama/v1/chat/completions
```

Here is my settings:
```
    {
      "model": "qwen2.5-coder:7b",
      "title": "My ollama",
      "provider": "openai",
      "useLegacyCompletionsEndpoint": false,
      "apiBase": "http://192.168.1.65:3000/ollama/v1",
      "apiKey": "sk-MY_API_KEY"
    }
```

Change it to provider: ollama, apiBase to http://192.168.1.65:3000/ollama, I got this error:
```
HTTP 400 Bad Request from http://192.168.1.65:3000/ollama/api/chat
```

**Reply by ProjectMoon** [reply]
What do the OpenWebUI logs say?

**Reply by jordanlambrecht** [reply]
Your provider type is still set to 'openai'

**Comment by flefevre** [comment]
Hello,
since i have opened this thread, and it is resolved with the following snipped of config.json file, I will close it.

`
{
    "models": [
        {
            "title": "mistralsmall-22b",
            "provider": "openai",
            "model": "mistralsmall-22b",
            "apiBase": "https://litellm.myhost.com/v1",
            "apiKey": "sk-xxxx"
        },
        {
            "title": "deepseek-coder-v2:16b",
            "provider": "openai",
            "model": "deepseek-coder-v2:16b",
            "apiBase": "https://ollama.myhost.com/v1",
            "apiKey": "toto"
        }
    ],
    "customCommands": [
            {
            "name": "test_fr",
            "prompt": "{{{ input }}}\n\nWrite a comprehensive set of unit tests for the selected code. It should setup, run tests that check for correctness including important edge cases, and teardown. Ensure that the tests are complete and sophisticated. Give the tests just as chat output, don't edit any file. Réponds en Français.",
            "description": "Write unit tests for highlighted code"
        },
        {
            "name": "build_readme_md_fr",
            "prompt": "{{{ input }}}\n\nBuild a readme.md fole for the application. Use badges and information about the licence as given by the user. Réponds en Français.",
            "description": "Build a readme.md code"
        }
    ],
    "tabAutocompleteModel": [
        {
            "title": "deepseek-coder-v2:16b",
            "provider": "openai",
            "model": "deepseek-coder-v2:16b",
            "apiBase": "https://ollama.myhost.com/v1",
            "apiKey": "toto"
        }
    ],
    "contextProviders": [
        {
            "name": "code",
            "params": {}
        },
        {
            "name": "docs",
            "params": {}
        },
        {
            "name": "diff",
            "params": {}
        },
        {
            "name": "terminal",
            "params": {}
        },
        {
            "name": "problems",
            "params": {}
        },
        {
            "name": "folder",
            "params": {}
        },
        {
            "name": "codebase",
            "params": {}
        }
    ],
    "slashCommands": [
        {
            "name": "edit",
            "description": "Edit selected code"
        },
        {
            "name": "commit",
            "description": "Generate a git commit message"
        }
    ],
    "embeddingsProvider": {
        "provider": "ollama",
        "model": "nomic-embed-text"
    },
    "allowAnonymousTelemetry": false
}

`

**Reply by ProjectMoon** [reply]
You might want to make a JSON config block.

```json
{ "like": "this" }
```

**Comment by teacher-svb** [comment]
I used the same tutorial and got it working... until two weeks ago! 

After an update from the Open WebUI instance from Truenas, I can no longer get any requests through. I only get an error:

`HTTP 405 Method Not Allowed from http://SERVER-IP:SERVER-PORT/v1/chat/completions` (IP and Port masked for obvious reasons)
